pkg/gardenlet/controller/vpaevictionrequirements/reconciler.go

```diff
@@ -0,0 +1,175 @@
+		return reconcile.Result{}, err
+	}
+
+	windowAnnotation := vpa.GetAnnotations()[constants.AnnotationShootMaintenanceWindow]
+	splitWindowAnnotation := strings.Split(windowAnnotation, ",")
+	if len(splitWindowAnnotation) != 2 {
+		err := fmt.Errorf("error during parsing the maintenance window from annotation. Value is not in format '<begin>,<end>': %q", windowAnnotation)
+		log.Error(err, "Error during reconciling for downscaling in maintenance only:")
+		return reconcile.Result{}, err
+	}
+	maintenanceTimeWindow, err := timewindow.ParseMaintenanceTimeWindow(splitWindowAnnotation[0], splitWindowAnnotation[1])
+	if err != nil {
+		log.Error(err, "Error during parsing the maintenance window from start and end time", "begin", splitWindowAnnotation[0], "end", splitWindowAnnotation[1])
+		return reconcile.Result{}, err
+	}
+
+	if isNowInMaintenanceTimeWindow := maintenanceTimeWindow.Contains(r.Clock.Now()); isNowInMaintenanceTimeWindow {
+		log.Info("Shoot is inside maintenance window, removing the EvictionRequirement to allow downscaling", "shoot-namespace", vpa.GetNamespace(), "maintenanceWindow", maintenanceTimeWindow)
+
+		if err := r.patchVPA(ctx, vpa, removeAllEvictionRequirements); err != nil {
+			return reconcile.Result{}, err
+		}
+
+		// requeue when the maintenance window ends, such that we can add the EvictionRequirement again
+		endTime := maintenanceTimeWindow.AdjustedEnd(r.Clock.Now())
+		requeueAfter := endTime.Sub(r.Clock.Now())
+		log.Info("Requeueing VPA", "namespace", vpa.Namespace, "name", vpa.Name, "requeueAfter", requeueAfter)
+		return reconcile.Result{RequeueAfter: requeueAfter}, nil
+	} else {
+		log.Info("Shoot is not inside maintenance window, adding EvictionRequirement to deny downscaling", "shoot-namespace", vpa.GetNamespace(), "maintenanceWindow", maintenanceTimeWindow)
+
+		if err := r.patchVPA(ctx, vpa, addDenyDownscalingEvictionRequirement); err != nil {
+			return reconcile.Result{}, err
+		}
+
+		// requeue when the next maintenance window begins, such that we can remove the EvictionRequirement
+		nextWindowBegin := maintenanceTimeWindow.AdjustedBegin(r.Clock.Now())
+		if nextWindowBegin.Before(r.Clock.Now()) {
+			nextWindowBegin = nextWindowBegin.AddDate(0, 0, 1)
+		}
+		requeueAfter := nextWindowBegin.Sub(r.Clock.Now())
+		log.Info("Requeueing VPA", "requeueAfter", requeueAfter)
+		return reconcile.Result{RequeueAfter: requeueAfter}, nil
+	}
+}
+
+func (r *Reconciler) reconcileVPAForDownscaleDisabled(ctx context.Context, vpa *vpaautoscalingv1.VerticalPodAutoscaler) (reconcile.Result, error) {
+	log := logf.FromContext(ctx)
+	log.Info("Adding EvictionRequirement for vpa to deny downscaling")
+
+	existing := vpa.DeepCopyObject()
+
+	addDenyDownscalingEvictionRequirement(vpa)
+
+	if equality.Semantic.DeepEqual(existing, vpa) {
+		return reconcile.Result{}, nil
+	}
+
+	if err := r.SeedClient.Update(ctx, vpa); err != nil {
+		return reconcile.Result{}, err
+	}
```

Comment:
Use `controllerutil.CreateOrUpdate` instead.

Reply:
We had a conversation about a similar thing above: Do you think there is a valid circumstance under which this controller should *create* VPA objects?

Reply:
Since this reconciler only reconciles existing VPA objects, it will never (re-)create them, so I guess using this utility function doesn't cause undesired side effects.

Reply:
done.